Cannot start Apache Spark on multi-host in Swarm mode

I simply cannot schedule instances of Apache Spark over the multi-host network in Swarm mode.

**Output of `docker version`:**

```
Client:
 Version:      1.12.0-rc4
 API version:  1.24
 Go version:   go1.6.2
 Git commit:   e4a0dbc
 Built:        Wed Jul 13 04:15:13 2016
 OS/Arch:      linux/amd64
 Experimental: true

Server:
 Version:      1.12.0-rc4
 API version:  1.24
 Go version:   go1.6.2
 Git commit:   e4a0dbc
 Built:        Wed Jul 13 04:15:13 2016
 OS/Arch:      linux/amd64
 Experimental: true
```


**Output of `docker info`:**

```
Containers: 3
 Running: 3
 Paused: 0
 Stopped: 0
Images: 1
Server Version: 1.12.0-rc4
Storage Driver: aufs
 Root Dir: /var/lib/docker/aufs
 Backing Filesystem: extfs
 Dirs: 12
 Dirperm1 Supported: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: host bridge overlay null
Swarm: active
 NodeID: 452huztay9dzxtehh50mzgf9o
 IsManager: Yes
 Managers: 1
 Nodes: 2
 CACertHash: sha256:2445358bb26a09d3b7cf67e0324b187fc6c0c4f71f0ae1413bd7b0838b26594f
Runtimes: runc
Default Runtime: runc
Security Options: apparmor seccomp
Kernel Version: 4.4.0-28-generic
Operating System: Ubuntu 16.04 LTS
OSType: linux
Architecture: x86_64
CPUs: 4
Total Memory: 7.797 GiB
Name: master
ID: QEVB:ADKU:BJFR:U7PB:NG2T:SBQH:B5NN:ZTNT:BLED:PE57:OY32:NXZQ
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
Experimental: true
Insecure Registries:
 127.0.0.0/8
WARNING: No swap limit support
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**
DigitalOcean


**Steps to reproduce the issue:**
1. Create 2 nodes `master` and `node1`. Do `docker swarm init` & `docker swarm join`
2. Create an overlay network, in my case: `docker network create --driver overlay spark`
3. Run `docker service create --name master --network spark -p 7077:7077 -p 8080:8080 gettyimages/spark:1.6.2-hadoop-2.6 bin/spark-class org.apache.spark.deploy.master.Master`
4. Run `docker service create --name worker --network spark -p 8081:8081 gettyimages/spark:1.6.2-hadoop-2.6 bin/spark-class org.apache.spark.deploy.worker.Worker spark://master:7077`
5. Scale worker to 2. You'll see it's trying to schedule a container on `node1` then failed and all 2 workers will be on master.

**Describe the results you received:**
Apache Spark instances were scheduled only on master nodes.

**Describe the results you expected:**
Apache Spark instances should be scheduled on both master and node1.
It is likely that containers on `node1` cannot resolve name `master` or cannot connect to the `master` IP correctly.

cc @mavenugo

Additional info
```
16/07/14 14:38:31 INFO worker.Worker: Registered signal handlers for [TERM, HUP, INT]
16/07/14 14:38:32 WARN util.NativeCodeLoader: Unable to load native-hadoop library for your platform... using builtin-java classes where applicable
16/07/14 14:38:32 INFO spark.SecurityManager: Changing view acls to: root
16/07/14 14:38:32 INFO spark.SecurityManager: Changing modify acls to: root
16/07/14 14:38:32 INFO spark.SecurityManager: SecurityManager: authentication disabled; ui acls disabled; users with view permissions: Set(root); users with modify permissions: Set(root)
16/07/14 14:38:33 INFO util.Utils: Successfully started service 'sparkWorker' on port 41470.
16/07/14 14:38:33 INFO worker.Worker: Starting Spark worker 10.0.0.5:41470 with 4 cores, 6.8 GB RAM
16/07/14 14:38:33 INFO worker.Worker: Running Spark version 1.6.2
16/07/14 14:38:33 INFO worker.Worker: Spark home: /usr/spark-1.6.2
16/07/14 14:38:33 INFO server.Server: jetty-8.y.z-SNAPSHOT
16/07/14 14:38:33 INFO server.AbstractConnector: Started SelectChannelConnector@0.0.0.0:8081
16/07/14 14:38:33 INFO util.Utils: Successfully started service 'WorkerUI' on port 8081.
16/07/14 14:38:33 INFO ui.WorkerWebUI: Started WorkerWebUI at http://10.0.0.5:8081
16/07/14 14:38:33 INFO worker.Worker: Connecting to master master:7077...
16/07/14 14:38:33 WARN worker.Worker: Failed to connect to master master:7077
java.io.IOException: Failed to connect to master:7077
        at org.apache.spark.network.client.TransportClientFactory.createClient(TransportClientFactory.java:216)
        at org.apache.spark.network.client.TransportClientFactory.createClient(TransportClientFactory.java:167)
        at org.apache.spark.rpc.netty.NettyRpcEnv.createClient(NettyRpcEnv.scala:200)
        at org.apache.spark.rpc.netty.Outbox$$anon$1.call(Outbox.scala:187)
        at org.apache.spark.rpc.netty.Outbox$$anon$1.call(Outbox.scala:183)
        at java.util.concurrent.FutureTask.run(FutureTask.java:266)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
        at java.lang.Thread.run(Thread.java:745)
Caused by: java.nio.channels.UnresolvedAddressException
        at sun.nio.ch.Net.checkAddress(Net.java:101)
        at sun.nio.ch.SocketChannelImpl.connect(SocketChannelImpl.java:622)
        at io.netty.channel.socket.nio.NioSocketChannel.doConnect(NioSocketChannel.java:209)
        at io.netty.channel.nio.AbstractNioChannel$AbstractNioUnsafe.connect(AbstractNioChannel.java:207)
        at io.netty.channel.DefaultChannelPipeline$HeadContext.connect(DefaultChannelPipeline.java:1097)
        at io.netty.channel.AbstractChannelHandlerContext.invokeConnect(AbstractChannelHandlerContext.java:471)
        at io.netty.channel.AbstractChannelHandlerContext.connect(AbstractChannelHandlerContext.java:456)
        at io.netty.channel.ChannelOutboundHandlerAdapter.connect(ChannelOutboundHandlerAdapter.java:47)
        at io.netty.channel.AbstractChannelHandlerContext.invokeConnect(AbstractChannelHandlerContext.java:471)
        at io.netty.channel.AbstractChannelHandlerContext.connect(AbstractChannelHandlerContext.java:456)
        at io.netty.channel.ChannelDuplexHandler.connect(ChannelDuplexHandler.java:50)
        at io.netty.channel.AbstractChannelHandlerContext.invokeConnect(AbstractChannelHandlerContext.java:471)
        at io.netty.channel.AbstractChannelHandlerContext.connect(AbstractChannelHandlerContext.java:456)
        at io.netty.channel.AbstractChannelHandlerContext.connect(AbstractChannelHandlerContext.java:438)
        at io.netty.channel.DefaultChannelPipeline.connect(DefaultChannelPipeline.java:908)
        at io.netty.channel.AbstractChannel.connect(AbstractChannel.java:203)
        at io.netty.bootstrap.Bootstrap$2.run(Bootstrap.java:166)
        at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:357)
        at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:357)
        at io.netty.util.concurrent.SingleThreadEventExecutor$2.run(SingleThreadEventExecutor.java:111)
        ... 1 more
```

It's clearly that containers on `node1` cannot resolve `master`.
Should I try `tasks.master` ?

I am running into a similar issue.  I'm not convinced it's a Spark-specific issue, but my Spark worker container is unable to connect to my Spark master container over an overlay network.

I'm running 1.12 on my Mac (Docker for Mac) in swarm mode (single manager node).

```
docker node ls
ID                           HOSTNAME  STATUS  AVAILABILITY  MANAGER STATUS
5ozf6nuyw1jgzujh7eyk9jzro *  moby      Ready   Active        Leader

docker info
Containers: 18
 Running: 4
 Paused: 0
 Stopped: 14
Images: 153
Server Version: 1.12.0
Storage Driver: aufs
 Root Dir: /var/lib/docker/aufs
 Backing Filesystem: extfs
 Dirs: 486
 Dirperm1 Supported: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: host overlay bridge null
Swarm: active
 NodeID: 5ozf6nuyw1jgzujh7eyk9jzro
 Is Manager: true
 ClusterID: 6vwuhafh0rjt00ncer0i7tpme
 Managers: 1
 Nodes: 1
 Orchestration:
  Task History Retention Limit: 5
 Raft:
  Snapshot interval: 10000
  Heartbeat tick: 1
  Election tick: 3
 Dispatcher:
  Heartbeat period: 5 seconds
 CA configuration:
  Expiry duration: 3 months
 Node Address: 192.168.65.2
Runtimes: runc
Default Runtime: runc
Security Options: seccomp
Kernel Version: 4.4.15-moby
Operating System: Alpine Linux v3.4
OSType: linux
Architecture: x86_64
CPUs: 4
Total Memory: 1.954 GiB
Name: moby
ID: DJB4:JAMU:H74J:SJSP:I372:2RQN:KFS5:RZ2B:NMBX:7KOH:Q6TX:USWQ
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): true
 File Descriptors: 72
 Goroutines: 182
 System Time: 2016-08-04T23:33:55.342277501Z
 EventsListeners: 5
Username: jtyberg
Registry: https://index.docker.io/v1/
Experimental: true
Insecure Registries:
 127.0.0.0/8
```

I'm deploying the following bundle:

```
cat spark.dab
{
  "Services": {
    "datanode1": {
      "Env": [
        "YARN_CONF_yarn_timeline___service_generic___application___history_enabled=true", 
        "YARN_CONF_yarn_resourcemanager_fs_state___store_uri=/rmstate", 
        "CORE_CONF_hadoop_http_staticuser_user=root", 
        "YARN_CONF_yarn_nodemanager_remote___app___log___dir=/app-logs", 
        "CORE_CONF_hadoop_proxyuser_hue_hosts=*", 
        "YARN_CONF_yarn_resourcemanager_recovery_enabled=true", 
        "YARN_CONF_yarn_timeline___service_hostname=historyserver", 
        "YARN_CONF_yarn_log___aggregation___enable=true", 
        "YARN_CONF_yarn_log_server_url=http://historyserver:8188/applicationhistory/logs/", 
        "YARN_CONF_yarn_resourcemanager_resource__tracker_address=resourcemanager:8031", 
        "YARN_CONF_yarn_resourcemanager_address=resourcemanager:8032", 
        "HDFS_CONF_dfs_permissions_enabled=false", 
        "CORE_CONF_fs_defaultFS=hdfs://namenode:8020", 
        "YARN_CONF_yarn_resourcemanager_system___metrics___publisher_enabled=true", 
        "CORE_CONF_hadoop_proxyuser_hue_groups=*", 
        "CORE_CONF_ipc_client_connect_max_retries=60", 
        "HDFS_CONF_dfs_webhdfs_enabled=true", 
        "YARN_CONF_yarn_timeline___service_enabled=true", 
        "YARN_CONF_yarn_resourcemanager_hostname=resourcemanager", 
        "YARN_CONF_yarn_resourcemanager_scheduler_address=resourcemanager:8030", 
        "YARN_CONF_yarn_resourcemanager_store_class=org.apache.hadoop.yarn.server.resourcemanager.recovery.FileSystemRMStateStore"
      ], 
      "Image": "bde2020/hadoop-datanode@sha256:e35ae0bf7a8de059d1082d62fefa3e5f8e03984a9d83134fa52d1a4bf4015ced", 
      "Networks": [
        "spark"
      ]
    }, 
    "master": {
      "Env": [
        "CORE_CONF_hadoop_http_staticuser_user=root", 
        "YARN_CONF_yarn_nodemanager_remote___app___log___dir=/app-logs", 
        "YARN_CONF_yarn_log_server_url=http://historyserver:8188/applicationhistory/logs/", 
        "HDFS_CONF_dfs_permissions_enabled=false", 
        "CORE_CONF_fs_defaultFS=hdfs://namenode:8020", 
        "YARN_CONF_yarn_resourcemanager_system___metrics___publisher_enabled=true", 
        "CORE_CONF_hadoop_proxyuser_hue_groups=*", 
        "YARN_CONF_yarn_log___aggregation___enable=true", 
        "SPARK_CONF_spark_eventLog_enabled=true", 
        "YARN_CONF_yarn_resourcemanager_scheduler_address=resourcemanager:8030", 
        "CORE_CONF_ipc_client_connect_max_retries=60", 
        "SPARK_MASTER_IP=master", 
        "YARN_CONF_yarn_resourcemanager_fs_state___store_uri=/rmstate", 
        "YARN_CONF_yarn_timeline___service_hostname=historyserver", 
        "YARN_CONF_yarn_resourcemanager_resource__tracker_address=resourcemanager:8031", 
        "YARN_CONF_yarn_resourcemanager_store_class=org.apache.hadoop.yarn.server.resourcemanager.recovery.FileSystemRMStateStore", 
        "YARN_CONF_yarn_resourcemanager_recovery_enabled=true", 
        "YARN_CONF_yarn_timeline___service_generic___application___history_enabled=true", 
        "YARN_CONF_yarn_timeline___service_enabled=true", 
        "YARN_CONF_yarn_resourcemanager_address=resourcemanager:8032", 
        "SPARK_CONF_spark_history_fs_logDirectory=hdfs://namenode:8020/spark-logs", 
        "SPARK_CONF_spark_eventLog_dir=hdfs://namenode:8020/spark-logs", 
        "CORE_CONF_hadoop_proxyuser_hue_hosts=*", 
        "HDFS_CONF_dfs_webhdfs_enabled=true", 
        "YARN_CONF_yarn_resourcemanager_hostname=resourcemanager"
      ], 
      "Image": "earthquakesan/hadoop-spark-master@sha256:941e26ac03d475734a5ebe698561febdaff10df818c72ba8d01d4b1436c59577", 
      "Networks": [
        "spark"
      ], 
      "Ports": [
        {
          "Port": 7077, 
          "Protocol": "tcp"
        }, 
        {
          "Port": 8080, 
          "Protocol": "tcp"
        }
      ]
    }, 
    "namenode": {
      "Env": [
        "CORE_CONF_hadoop_http_staticuser_user=root", 
        "YARN_CONF_yarn_nodemanager_remote___app___log___dir=/app-logs", 
        "YARN_CONF_yarn_log_server_url=http://historyserver:8188/applicationhistory/logs/", 
        "HDFS_CONF_dfs_permissions_enabled=false", 
        "CORE_CONF_fs_defaultFS=hdfs://namenode:8020", 
        "YARN_CONF_yarn_resourcemanager_system___metrics___publisher_enabled=true", 
        "CORE_CONF_hadoop_proxyuser_hue_groups=*", 
        "YARN_CONF_yarn_log___aggregation___enable=true", 
        "YARN_CONF_yarn_resourcemanager_scheduler_address=resourcemanager:8030", 
        "CORE_CONF_ipc_client_connect_max_retries=60", 
        "YARN_CONF_yarn_resourcemanager_fs_state___store_uri=/rmstate", 
        "YARN_CONF_yarn_timeline___service_hostname=historyserver", 
        "YARN_CONF_yarn_resourcemanager_resource__tracker_address=resourcemanager:8031", 
        "YARN_CONF_yarn_resourcemanager_store_class=org.apache.hadoop.yarn.server.resourcemanager.recovery.FileSystemRMStateStore", 
        "YARN_CONF_yarn_resourcemanager_recovery_enabled=true", 
        "YARN_CONF_yarn_timeline___service_generic___application___history_enabled=true", 
        "CLUSTER_NAME=test", 
        "YARN_CONF_yarn_timeline___service_enabled=true", 
        "YARN_CONF_yarn_resourcemanager_address=resourcemanager:8032", 
        "CORE_CONF_hadoop_proxyuser_hue_hosts=*", 
        "HDFS_CONF_dfs_webhdfs_enabled=true", 
        "YARN_CONF_yarn_resourcemanager_hostname=resourcemanager"
      ], 
      "Image": "bde2020/hadoop-namenode@sha256:c07c915f7e90f8ae312083afed35e79f74c40126f15cc516d97b4d86be62a917", 
      "Networks": [
        "spark"
      ], 
      "Ports": [
        {
          "Port": 50070, 
          "Protocol": "tcp"
        }, 
        {
          "Port": 8020, 
          "Protocol": "tcp"
        }
      ]
    }, 
    "worker": {
      "Env": [
        "CORE_CONF_hadoop_http_staticuser_user=root", 
        "YARN_CONF_yarn_nodemanager_remote___app___log___dir=/app-logs", 
        "YARN_CONF_yarn_log_server_url=http://historyserver:8188/applicationhistory/logs/", 
        "HDFS_CONF_dfs_permissions_enabled=false", 
        "CORE_CONF_fs_defaultFS=hdfs://namenode:8020", 
        "YARN_CONF_yarn_resourcemanager_system___metrics___publisher_enabled=true", 
        "CORE_CONF_hadoop_proxyuser_hue_groups=*", 
        "YARN_CONF_yarn_log___aggregation___enable=true", 
        "SPARK_CONF_spark_eventLog_enabled=true", 
        "YARN_CONF_yarn_resourcemanager_scheduler_address=resourcemanager:8030", 
        "CORE_CONF_ipc_client_connect_max_retries=60", 
        "YARN_CONF_yarn_resourcemanager_fs_state___store_uri=/rmstate", 
        "YARN_CONF_yarn_timeline___service_hostname=historyserver", 
        "YARN_CONF_yarn_resourcemanager_resource__tracker_address=resourcemanager:8031", 
        "YARN_CONF_yarn_resourcemanager_store_class=org.apache.hadoop.yarn.server.resourcemanager.recovery.FileSystemRMStateStore", 
        "YARN_CONF_yarn_resourcemanager_recovery_enabled=true", 
        "YARN_CONF_yarn_timeline___service_generic___application___history_enabled=true", 
        "YARN_CONF_yarn_timeline___service_enabled=true", 
        "SPARK_MASTER_URL=spark://master:7077", 
        "YARN_CONF_yarn_resourcemanager_address=resourcemanager:8032", 
        "SPARK_CONF_spark_history_fs_logDirectory=hdfs://namenode:8020/spark-logs", 
        "SPARK_CONF_spark_eventLog_dir=hdfs://namenode:8020/spark-logs", 
        "CORE_CONF_hadoop_proxyuser_hue_hosts=*", 
        "HDFS_CONF_dfs_webhdfs_enabled=true", 
        "YARN_CONF_yarn_resourcemanager_hostname=resourcemanager"
      ], 
      "Image": "earthquakesan/hadoop-spark-worker@sha256:e0a3ac8d6c19705838fe10ef5d3afc4bc41914ead28331e0804f4743e48dc93a", 
      "Networks": [
        "spark"
      ]
    }
  }, 
  "Version": "0.1"
}
```

My Spark master is exposing ports `7077` and `8080`.

Docker creates the `spark_spark` overlay network and successfully starts the services.

```
docker deploy spark
Loading bundle from spark.dab
Creating network spark_spark
Creating service spark_worker
Creating service spark_datanode1
Creating service spark_master
Creating service spark_namenode

docker network ls
NETWORK ID          NAME                    DRIVER              SCOPE
3d54b6e719da        bridge                  bridge              local               
38b675c2e238        docker_gwbridge         bridge              local               
d68274d282dc        host                    host                local               
dav49l3pta6u        ingress                 overlay             swarm               
88f346476511        none                    null                local               
cayf5zg9yaov        spark_spark             overlay             swarm               
```

Note the IP address of the Spark master container is `10.0.0.5` on the `spark_spark` network.

```
docker network inspect spark_spark
[
    {
        "Name": "spark_spark",
        "Id": "cayf5zg9yaov9nv0jsitmq8fg",
        "Scope": "swarm",
        "Driver": "overlay",
        "EnableIPv6": false,
        "IPAM": {
            "Driver": "default",
            "Options": null,
            "Config": [
                {
                    "Subnet": "10.0.0.0/24",
                    "Gateway": "10.0.0.1"
                }
            ]
        },
        "Internal": false,
        "Containers": {
            "68635acfca29e359c0ae9eef60766b58b1b7b8daf020d3a7351a257d9b513238": {
                "Name": "spark_namenode.1.54i7jygy38hyow3sj07791guv",
                "EndpointID": "d1379f120390c293f41abfaa1ddef74984f2d982f6c117b93bb02531d8c7cfd6",
                "MacAddress": "02:42:0a:00:00:07",
                "IPv4Address": "10.0.0.7/24",
                "IPv6Address": ""
            },
            "a67525c16046db535b3cbce557faf109f7bd86df90226b67c48931b9a0401b21": {
                "Name": "spark_datanode1.1.1d1w8vnxzof44xbf05ckxbfq2",
                "EndpointID": "e80df47fe27fea9159d0525fe81362a1cfc756bb24d9c422a492037e1cc8cf35",
                "MacAddress": "02:42:0a:00:00:09",
                "IPv4Address": "10.0.0.9/24",
                "IPv6Address": ""
            },
            "c26813df3eb3ce7eb98ed6a5217ede0d5f628d67a7f9483070bfa188983e1b1f": {
                "Name": "spark_master.1.39129b3zeqdm80rmb1lamz42j",
                "EndpointID": "53dd0274a4482dbcac3ecd61b98ed5b170bd9899859a7d60de4af8117ea48b75",
                "MacAddress": "02:42:0a:00:00:05",
                "IPv4Address": "10.0.0.5/24",
                "IPv6Address": ""
            },
            "e301de3ee41b1a1a763620cc8c88a6279d16a4ff5de7d00be5e172703470a50e": {
                "Name": "spark_worker.1.9n4d2iahfgataj9k9k40oahpl",
                "EndpointID": "703dbb08d73c105d3ed2af0d7db342620f9c3e3c7a046c41d88adfaee62445a1",
                "MacAddress": "02:42:0a:00:00:03",
                "IPv4Address": "10.0.0.3/24",
                "IPv6Address": ""
            }
        },
        "Options": {
            "com.docker.network.driver.overlay.vxlanid_list": "257"
        },
        "Labels": {
            "com.docker.stack.namespace": "spark"
        }
    }
]
```

On the swarm containers, `master` resolves to `10.0.0.6`.  My spark master container is listening on `10.0.0.6:7077`.

```
docker exec -it spark_master.1.39129b3zeqdm80rmb1lamz42j netstat -tupln
Active Internet connections (only servers)
Proto Recv-Q Send-Q Local Address           Foreign Address         State       PID/Program name
tcp        0      0 127.0.0.11:33807        0.0.0.0:*               LISTEN      -               
tcp6       0      0 10.0.0.6:7077           :::*                    LISTEN      25/java         
tcp6       0      0 :::8080                 :::*                    LISTEN      25/java         
tcp6       0      0 10.0.0.6:6066           :::*                    LISTEN      25/java         
udp        0      0 127.0.0.11:46310        0.0.0.0:*                           -               
```

My spark worker container can't connect to the `master:7077`.  It resolves `master` to `10.0.0.6`, but it can't seem to route to it over the `spark_spark` network.

```
docker logs spark_worker.1.9n4d2iahfgataj9k9k40oahpl
16/08/04 23:29:38 WARN worker.Worker: Failed to connect to master master:7077
java.io.IOException: Failed to connect to master/10.0.0.6:7077
	at org.apache.spark.network.client.TransportClientFactory.createClient(TransportClientFactory.java:216)
	at org.apache.spark.network.client.TransportClientFactory.createClient(TransportClientFactory.java:167)
	at org.apache.spark.rpc.netty.NettyRpcEnv.createClient(NettyRpcEnv.scala:200)
	at org.apache.spark.rpc.netty.Outbox$$anon$1.call(Outbox.scala:187)
	at org.apache.spark.rpc.netty.Outbox$$anon$1.call(Outbox.scala:183)
	at java.util.concurrent.FutureTask.run(FutureTask.java:262)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)
	at java.lang.Thread.run(Thread.java:745)
Caused by: java.net.ConnectException: Connection refused: master/10.0.0.6:7077
	at sun.nio.ch.SocketChannelImpl.checkConnect(Native Method)
	at sun.nio.ch.SocketChannelImpl.finishConnect(SocketChannelImpl.java:744)
	at io.netty.channel.socket.nio.NioSocketChannel.doFinishConnect(NioSocketChannel.java:224)
	at io.netty.channel.nio.AbstractNioChannel$AbstractNioUnsafe.finishConnect(AbstractNioChannel.java:289)
	at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:528)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:468)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:382)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:354)
	at io.netty.util.concurrent.SingleThreadEventExecutor$2.run(SingleThreadEventExecutor.java:111)
	... 1 more
```

However, I CAN get to the spark master web UI at `master:8080` from the worker.  I suspect it's somehow getting there through the ingress network.

```
docker exec -it spark_worker.1.9n4d2iahfgataj9k9k40oahpl curl -v http://master:8080 -o temp
* About to connect() to master port 8080 (#0)
*   Trying 10.0.0.6...
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0* connected
* Connected to master (10.0.0.6) port 8080 (#0)
> GET / HTTP/1.1
> User-Agent: curl/7.26.0
> Host: master:8080
> Accept: */*
> 
* additional stuff not fine transfer.c:1037: 0 0
* HTTP 1.1 or later with persistent connection, pipelining supported
< HTTP/1.1 200 OK
< Content-Type: text/html;charset=UTF-8
< Cache-Control: no-cache, no-store, must-revalidate
< X-Frame-Options: SAMEORIGIN
< Content-Length: 3989
< Server: Jetty(8.y.z-SNAPSHOT)
< 
{ [data not shown]
100  3989  100  3989    0     0   169k      0 --:--:-- --:--:-- --:--:--  194k
* Connection #0 to host master left intact
* Closing connection #0
```

In summary, 

* IP address of `master` as listed in `docker network inspect` is `10.0.0.5`
* IP address that `master` resolves to within containers is `10.0.0.6`
* `master` container is listening on `10.0.0.6:7077` and `:::8080`
* `worker` container can connect to `master:8080`
* `worker` container CANNOT connect to `master:7077`

--- a/daemon/daemon_unix.go
+++ b/daemon/daemon_unix.go
@@ -41,6 +41,7 @@ import (
 	rsystem "github.com/opencontainers/runc/libcontainer/system"
 	"github.com/opencontainers/runc/libcontainer/user"
 	specs "github.com/opencontainers/runtime-spec/specs-go"
+	"github.com/pkg/errors"
 	"github.com/vishvananda/netlink"
 )
 
@@ -710,13 +711,30 @@ func initBridgeDriver(controller libnetwork.NetworkController, config *Config) e
 
 	ipamV4Conf = &libnetwork.IpamConf{AuxAddresses: make(map[string]string)}
 
-	nw, nw6List, err := netutils.ElectInterfaceAddresses(bridgeName)
-	if err == nil {
-		ipamV4Conf.PreferredPool = lntypes.GetIPNetCanonical(nw).String()
-		hip, _ := lntypes.GetHostPartIP(nw.IP, nw.Mask)
-		if hip.IsGlobalUnicast() {
-			ipamV4Conf.Gateway = nw.IP.String()
+	nwList, nw6List, err := netutils.ElectInterfaceAddresses(bridgeName)
+	if err != nil {
+		return errors.Wrap(err, "list bridge addresses failed")
+	}
+
+	nw := nwList[0]
+	if len(nwList) > 1 && config.bridgeConfig.FixedCIDR != "" {
+		_, fCIDR, err := net.ParseCIDR(config.bridgeConfig.FixedCIDR)
+		if err != nil {
+			return errors.Wrap(err, "parse CIDR failed")
 		}
+		// Iterate through in case there are multiple addresses for the bridge
+		for _, entry := range nwList {
+			if fCIDR.Contains(entry.IP) {
+				nw = entry
+				break
+			}
+		}
+	}
+
+	ipamV4Conf.PreferredPool = lntypes.GetIPNetCanonical(nw).String()
+	hip, _ := lntypes.GetHostPartIP(nw.IP, nw.Mask)
+	if hip.IsGlobalUnicast() {
+		ipamV4Conf.Gateway = nw.IP.String()
 	}
 
 	if config.bridgeConfig.IP != "" {

--- a/hack/vendor.sh
+++ b/hack/vendor.sh
@@ -70,7 +70,7 @@ clone git github.com/RackSec/srslog 365bf33cd9acc21ae1c355209865f17228ca534e
 clone git github.com/imdario/mergo 0.2.1
 
 #get libnetwork packages
-clone git github.com/docker/libnetwork 9fbb4ecbb45af655c4ac3c2f3a849b2294cb447a
+clone git github.com/docker/libnetwork f4338b6f1085ccfe5972e655cca8a1d15d73439d
 clone git github.com/docker/go-events 18b43f1bc85d9cdd42c05a6cd2d444c7a200a894
 clone git github.com/armon/go-radix e39d623f12e8e41c7b5529e9a9dd67a1e2261f80
 clone git github.com/armon/go-metrics eb0af217e5e9747e41dd5303755356b62d28e3ec

